File: Demos/Datasnap/Server/proxy/java_android/com/embarcadero/javaandroid/TJSONObject.java
//*******************************************************
//
//               Delphi DataSnap Framework
//
// Copyright(c) 1995-2011 Embarcadero Technologies, Inc.
//
//*******************************************************

package com.embarcadero.javaandroid;

import java.util.LinkedList;
import java.util.List;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * 
 * Implements a JSON object.
 * 
 */

public class TJSONObject extends TJSONValue {

	protected List<TJSONPair> Elements;

	@Override
	public JSONValueType getJsonValueType() {
		return JSONValueType.JSONObject;
	}

	protected List<TJSONPair> buildElements(JSONObject o) {
		try {
			String pname;
			List<TJSONPair> res = new LinkedList<TJSONPair>();
			JSONArray keys = o.names();
			for (int i = 0; i < keys.length(); i++) {
				pname = keys.getString(i);
				Object obj = o.get(pname);
				if (obj == JSONObject.NULL) {
					res.add(new TJSONPair(pname, new TJSONNull()));
				} else if (obj instanceof String) {
					res.add(new TJSONPair(pname, ((String) obj)));
				} else if (obj instanceof Double) {
					res.add(new TJSONPair(pname, new TJSONNumber(((Double) obj)
							.doubleValue())));
				} else if (obj instanceof Integer) {
					res.add(new TJSONPair(pname, new TJSONNumber(
							((Integer) obj).intValue())));
				} else if (obj instanceof JSONArray) {
					res.add(new TJSONPair(pname, new TJSONArray(
							((JSONArray) obj))));
				} else if (obj instanceof JSONObject) {
					res.add(new TJSONPair(pname, new TJSONObject(
							((JSONObject) obj))));
				} else if (obj instanceof Boolean) {
					if (((Boolean) obj).booleanValue())
						res.add(new TJSONPair(pname, new TJSONTrue()));
					else
						res.add(new TJSONPair(pname, new TJSONFalse()));
				}
			}
			return res;
		} catch (JSONException ex) {
			return null;
		}

	}

	protected JSONObject asJSONObject() {
		try {
			JSONObject j = new JSONObject();
			for (TJSONPair pair : Elements) {
				switch (pair.value.getJsonValueType()) {
				case JSONObject: {
					j.put(pair.name, ((TJSONObject) pair.value).asJSONObject());
					break;
				}
				case JSONArray: {
					j.put(pair.name, ((TJSONArray) pair.value).asJSONArray());
					break;
				}
				case JSONString: {
					j.put(pair.name, ((TJSONString) pair.value).getValue());
					break;
				}
				case JSONNumber: {
					j.put(pair.name, ((TJSONNumber) pair.value).getValue());
					break;
				}
				case JSONTrue: {
					j.put(pair.name, true);
					break;
				}
				case JSONFalse: {
					j.put(pair.name, false);
					break;
				}
				case JSONNull: {
					j.put(pair.name, JSONObject.NULL);
					break;
				}
				}
			}
			return j;
		} catch (JSONException ex) {
			return null;
		}
	}

	/**
	 * Class constructor, initialized the internal list
	 */
	public TJSONObject() {
		super();
		Elements = new LinkedList<TJSONPair>();
	}

	/**
	 * Parse the passed String into a new TJSONObject
	 * 
	 * @param value
	 * @return
	 */
	public static TJSONObject Parse(String value) {
		try {
			JSONObject o = new JSONObject(value);
			return new TJSONObject(o);
		} catch (JSONException e) {
			return null;
		}
	}

	/**
	 * Initialized the instance with a JSONObject;
	 * 
	 * @param json
	 */
	public TJSONObject(JSONObject json) {
		super();
		Elements = buildElements(json);
	}

	/**
	 * Class constructor, create the internal list and add the passed
	 * {@link TJSONPair}
	 * 
	 * @param pair
	 */
	public TJSONObject(TJSONPair pair) {
		super();
		Elements = new LinkedList<TJSONPair>();
		addPairs(pair);
	}

	@Override
	public String toString() {
		return asJSONObject().toString();
	}

	@Override
	public Object getInternalObject() {
		return asJSONObject();
	}

	/**
	 * Returns a String value by the name
	 * 
	 * @param name
	 * @return
	 */
	public String getString(String name) {
		TJSONPair p;
		return ((p = get(name)) == null) ? null : ((TJSONString) p.value)
				.getValue();
	}

	/**
	 * Returns a Boolean value by the name
	 * 
	 * @param name
	 * @return
	 */
	public Boolean getBoolean(String name) {
		TJSONPair p = get(name);
		if (p == null)
			return null;
		if (p.value instanceof TJSONTrue)
			return true;
		else
			return false;
	}

	/**
	 * Returns a Double value by the name
	 * 
	 * @param name
	 * @return
	 */
	public Double getDouble(String name) {
		TJSONPair p;
		return ((p = get(name)) == null) ? null : ((TJSONNumber) p.value)
				.getValue();
	}

	/**
	 * Returns a TJSONObject value by the name
	 * 
	 * @param name
	 * @return
	 */
	public TJSONObject getJSONObject(String name) {
		TJSONPair p;
		return ((p = get(name)) == null) ? null : ((TJSONObject) (p.value));
	}

	/**
	 * Returns a Integer value by the name
	 * 
	 * @param name
	 * @return
	 */
	public Integer getInt(String name) {
		TJSONPair p;
		return ((p = get(name)) == null) ? null : ((TJSONNumber) p.value)
				.getValue().intValue();
	}

	/**
	 * Returns a {@link TJSONArray} value by the name
	 * 
	 * @param name
	 * @return
	 */
	public TJSONArray getJSONArray(String name) {
		TJSONPair p;
		return ((p = get(name)) == null) ? null : ((TJSONArray) p.value);
	}

	/**
	 * Detects if it has a key with the name specified.
	 * 
	 * @param name
	 * @return
	 */
	public boolean has(String name) {
		return get(name) != null;
	}

	/**
	 * Gets a {@link TJSONPair} from a name
	 * 
	 * @param name
	 * @return
	 */
	public TJSONPair get(String name) {
		for (TJSONPair v : Elements) {
			if (v.name.equals(name))
				return v;
		}
		return null;
	}

	/**
	 * Adds {@link TJSONPair}
	 * 
	 * @param pair
	 * @return
	 */
	public TJSONObject addPairs(TJSONPair pair) {
		Elements.add(pair);
		return this;
	}

	/**
	 * Adds a pair with name and a {@link TJSONValue}
	 * 
	 * @param name
	 * @param value
	 * @return
	 */
	public TJSONObject addPairs(String name, TJSONValue value) {
		return addPairs(new TJSONPair(name, value));
	}

	/**
	 * Adds a pair with name and an int value
	 * 
	 * @param name
	 * @param value
	 * @return
	 */
	public TJSONObject addPairs(String name, int value) {
		return addPairs(name, new TJSONNumber(value));
	}

	/**
	 * Adds a pair with name and a String value
	 * 
	 * @param name
	 * @param value
	 * @return
	 */
	public TJSONObject addPairs(String name, String value) {
		return addPairs(name, new TJSONString(value));
	}

	/**
	 * Adds a pair with name and a long value
	 * 
	 * @param name
	 * @param value
	 * @return
	 */
	public TJSONObject addPairs(String name, long value) {
		return addPairs(name, new TJSONNumber(value));
	}

	/**
	 * Adds a pair with name and a double value
	 * 
	 * @param name
	 * @param value
	 * @return
	 */
	public TJSONObject addPairs(String name, double value) {
		return addPairs(name, new TJSONNumber(value));
	}

	/**
	 * Adds a pair with name and a boolean value
	 * 
	 * @param name
	 * @param value
	 * @return
	 */
	public TJSONObject addPairs(String name, boolean value) {
		if (value)
			return addPairs(name, new TJSONTrue());
		else
			return addPairs(name, new TJSONFalse());
	}
}

File: Demos/Datasnap/Server/proxy/objectivec_ios42/Proxy/TJSONObject.m
//*******************************************************
//
//               Delphi DataSnap Framework
//
// Copyright(c) 1995-2011 Embarcadero Technologies, Inc.
//
//*******************************************************


#import "TJSONObject.h"
#import "TJSONNull.h"
#import "TJSONString.h"
#import "TJSONNumber.h"
#import "TJSONArray.h"
#import "TJSONTrue.h"
#import "TJSONFalse.h"
#import "SBJSON.h"


#import "DBXJsonTools.h"


@implementation TJSONObject
- (NSDictionary * ) JSONObjectFromString:(NSString *)json_string{
	SBJsonParser * parser = [[SBJsonParser alloc] init];
	@try {
		return [parser objectWithString:json_string];
	}
	@finally {
		[parser release];
	}
	
	
}
-(NSString * )JSONObjectToString:(NSDictionary * )json{
	SBJsonWriter * parser = [[SBJsonWriter alloc] init];
	@try {
		return [parser stringWithObject:json];
	}
	@finally {
		[parser release];
	}
}



-(TJSONPairList *) buildElements:(NSDictionary * )o {
	@try {
		
		TJSONPairList* res =[[[TJSONPairList alloc] init] autorelease]  ;
		for (NSString * pname in o) {
			id obj = [o objectForKey:pname];
			if ([obj isKindOfClass:[NSNull class]]) {
				[res addPair: [TJSONPair JSONPairWithName:pname andValue:[[[TJSONNull alloc]init]autorelease]]];
			} else if ([obj isKindOfClass:[NSString class]]) {
				[res addPair: [TJSONPair JSONPairWithName:pname andValue:[[[TJSONString alloc]initWithString:(NSString *)obj]autorelease]]];
			} else if ([obj isKindOfClass:[NSArray class]]) {
				[res addPair: [TJSONPair JSONPairWithName:pname andValue:[[[TJSONArray alloc]initWithJSONArray:obj]autorelease]]];
			}else if ([obj isKindOfClass:[TJSONObject class]]) {
				[res addPair:[TJSONPair JSONPairWithName:pname andValue:obj]];
			}else if ([obj isKindOfClass:[NSDictionary class]]) {
				[res addPair: [TJSONPair JSONPairWithName:pname andValue:[[[TJSONObject alloc]initWithJSONObject:obj]autorelease]]];
			} else if ([obj isKindOfClass:[NSNumber class]]) {
				if (strcmp([obj objCType], @encode(BOOL)) == 0){
					if ([obj boolValue]) {
						[res addPair: [TJSONPair JSONPairWithName:pname andValue:[[[TJSONTrue alloc]init]autorelease]]];}
					else{
						[res addPair: [TJSONPair JSONPairWithName:pname andValue:[[[TJSONFalse alloc]init]autorelease]]];}
				}else{
					[res addPair: [TJSONPair JSONPairWithName:pname andValue:[[[TJSONNumber alloc]initWithNumber:obj]autorelease]]];}
			}			
			
		}
		return res;		
	} @catch (NSException *ex) {
		return nil;
		@throw ex;
	}
	
}

-(id) init{
	self = [super init];
	if (self) {
	  elements = [[TJSONPairList alloc]init];
	}
	
	return self;
}
-(id) initWithJSONString:(NSString *) stringJson{
	
	self = [self init];
	if (self) {
		[elements release];
		elements  = [[self buildElements: [self JSONObjectFromString:stringJson]] retain];

	}
	return self;
}
-(id) initWithJSONObject:(NSDictionary *) json{
	self = [self init];
	if (self) {
		[elements release];
		elements  = [[self buildElements: json ] retain];

	}
	return self;
	
}
-(NSString *) asJSONString{
	return [self JSONObjectToString:[self asJSONObject]];  
}

-(NSString *) toString {
	return [self asJSONString];
}
-(NSDictionary *)asJSONObject {
	return [elements asJSONObject];
}
-(NSArray *) asJSONArray {
	return nil;
}

-(id) getInternalObject{
	return [elements asJSONObject ];
}
-(void) dealloc {
	[elements release];
	[super dealloc];
}
-(JSONValueType) getJSONValueType{
	return JSONObject;
}
-(id) addPairs:(TJSONPair *)value{
	[elements addPair:value];
	return self;
}
-(id) addPairs: (NSString*) name withJSONValue:(TJSONValue *) value{
	[elements addPair:[TJSONPair JSONPairWithName:name andValue:value]];
	return self;
}
-(id) addPairs:(NSString *) name withString:(NSString *)value{
	[elements addPair:[TJSONPair JSONPairWithName:name andValue:
					   [TJSONString JSONStringWithString:value]] ] ;
	
	return self;
	
}

-(id) addPairs:(NSString *)name withInt:(int)value{
	
	[elements addPair:[TJSONPair JSONPairWithName:name andValue:
					   [TJSONNumber jsonNumberWithNumber:[NSNumber numberWithInt:value]]]];
	
	return self;
}
-(id) addPairs:(NSString *)name withLong:(long)value{
	[elements addPair:[TJSONPair JSONPairWithName:name andValue:
					   [TJSONNumber jsonNumberWithNumber:[NSNumber numberWithLong:value]]]];
	
	return self;
}
-(id) addPairs:(NSString *)name withDouble:(double)value{
	[elements addPair:[TJSONPair JSONPairWithName:name andValue:
					   [TJSONNumber jsonNumberWithNumber:[NSNumber numberWithDouble:value]]]];
	
	return self;
}

-(id) addPairs:(NSString *)name withBool:(BOOL)value{
if (value) {
	[elements addPair:[TJSONPair JSONPairWithName:name andValue:[[[TJSONTrue alloc]init]autorelease]]];
}else{
	[elements addPair:[TJSONPair JSONPairWithName:name andValue:[[[TJSONFalse alloc]init]autorelease]]];
}
	return self;
}


-(NSString *) getStringForKey:(NSString *)name{
	TJSONPair* p =  [elements PairByName:name];
   	return [(TJSONString *)p.value getValue];
}
-(TJSONArray *) getJSONArrayForKey:(NSString *)name{
	TJSONPair* p =  [elements PairByName:name];
	return (TJSONArray *) p.value ;
}
-(TJSONObject *) getJSONObjectForKey:(NSString *)name{
	TJSONPair* p =  [elements PairByName:name];
	return (TJSONObject *) p.value ;
}
-(int) getIntForKey:(NSString * ) name{
	TJSONPair* p =  [elements PairByName:name];
	return [[p.value getInternalObject]intValue] ;
}
-(long) getLongForKey:(NSString * ) name{
	TJSONPair* p =  [elements PairByName:name];
	return [[p.value getInternalObject]longValue] ;
}
-(double) getDoubleForKey:(NSString * ) name{
	TJSONPair* p =  [elements PairByName:name];
	return [[p.value getInternalObject]doubleValue] ;
}
-(BOOL) getBoolForKey:(NSString * ) name{
	TJSONPair* p =  [elements PairByName:name];
	if ([p.value isKindOfClass:[TJSONTrue class]]) {
        return YES;
	}
	return NO;
}
-(TJSONPair *)getPairForKey:(NSString *)name{
  return[elements PairByName:name];
}
-(BOOL) hasKey:(NSString *)name{
	return !([elements PairByName:name] == nil);
}
-(int) count{
	return [elements count];
}

@end
@implementation TJSONObject(jsonobjectCreations)
+(id)parse:(NSString *) value{
	return [[[TJSONObject alloc]initWithJSONString:value]autorelease];
}
@end
